sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,1003 @@
+        check_scalar(self.max_iter, 'max_iter', int, 1)
+        check_scalar(self.max_constraints, 'max_constraints', int, 1)
+        check_scalar(self.max_corrections, 'max_corrections', int, 1)
+        check_scalar(self.n_jobs, 'n_jobs', int, -1)
+
+        check_scalar(self.tol, 'tol', float, 0.)
+
+        check_scalar(self.use_pca, 'use_pca', bool)
+        check_scalar(self.use_sparse, 'use_sparse', bool)
+        check_scalar(self.verbose, 'verbose', int, 0)
+
+        if self.callback is not None:
+            if not callable(self.callback):
+                raise ValueError('callback is not callable.')
+
+        # Check linear transformation dimensions
+        if self.L is not None:
+            check_array(self.L)
+            if len(self.L[0]) != n_features:
+                raise ValueError('Transformation input dimensionality ({}) '
+                                 'must match the inputs dimensionality ({}).'
+                                 .format(len(self.L[0]), n_features))
+
+            if len(self.L) > len(self.L[0]):
+                raise ValueError('Transformation output dimensionality ({}) '
+                                 'cannot be greater than the '
+                                 'transformation input dimensionality ({}).'.
+                                 format(len(self.L), len(self.L[0])))
+
+        # Check preferred output dimensionality
+        if self.n_features_out is not None:
+            if self.L is not None:
+                if self.n_features_out != len(self.L):
+                    raise ValueError('Preferred outputs dimensionality ({}) '
+                                     'does not match the given linear '
+                                     'transformation {}!'.format(
+                                        self.n_features_out, len(self.L)))
+
+            elif self.n_features_out > n_features:
+                raise ValueError('Preferred outputs dimensionality ({}) '
+                                 'cannot be greater than the given data '
+                                 'dimensionality {}!'.format(
+                                    self.n_features_out, n_features))
+
+        # Check preferred number of neighbors
+        min_non_singleton_size = class_sizes[~is_class_singleton].min()
+        if self.n_neighbors >= min_non_singleton_size:
+            warn("n_neighbors (={}) is not less than the number of samples in "
+                 "the smallest non-singleton class (={}). n_neighbors will be "
+                 "set to (min_non_singleton_size - 1) for estimation."
+                 .format(self.n_neighbors, min_non_singleton_size))
+
+        self.n_neighbors_ = min(self.n_neighbors, min_non_singleton_size - 1)
+        # TODO: Notify superclass KNeighborsClassifier that n_neighbors
+        # might have changed to n_neighbors_
+        # super(LargeMarginNearestNeighbor, self).set_params(
+        # n_neighbors=self.n_neighbors_)
```

Comment:
From a use-case point of view I think this would make sense for users who are primarily interested in learning a metric to transform their data (without K-NN classification). In this case LargeMarginNearestNeighbor would not be a classifier anymore but a transformer.